### ISSUE TITLE
Optimize Packet

### ISSUE THREAD

**Issue by aerth** [mmogo/mmo#17]
The packets should be as small as possible and not contain any empty fields.

Update packets really don't need Request fields in the message, and vice versa.

We should consider an encoding that uses **struct field tags** so that we may `omitempty` or `opt`

Also consider a message envelope that details which sort of message is contained, or using a **type switch** after/inside `GetMessage`

```go
        msg, err := shared.GetMessage(conn)
        if err != nil {
                // disconnected
        }
        switch msg.(type) {
        default: // handle invalid packet
        case Ping:
                // handle pong
        case Update:
                // handle update
        case Request:
                // handle request
        }

```
From https://golang.org/pkg/encoding/gob/
> The field value is sent using the standard gob encoding for its type, recursively. If a field has the zero value for its type (except for arrays; see above), it is omitted from the transmission.

It doesn't seem nil pointers are completely omitted when using gob encoding.

Here is the ping, which is only Sent (`time.Time`) with nil Update and nil Request: 
```
[udp sum ok] UDP, length 701
......H.$.......E....F@.@......,...........)...QQ. .%o..T...".........8......Message........Sent......Request......Update.............Time......L......Request........ConnectRequest......MoveRequest......SpeakRequest......#......ConnectRequest........ID.....4......MoveRequest.......      Direction.....Created......#......SpeakRequest........Text.....^......Update........PlayerMoved......PlayerSpoke.....
WorldState......PlayerDisconnected......B......PlayerMoved........ID.....NewPosition......RequestTime.............Vec........X.....Y.....)......PlayerSpoke........ID.....Text.....%.....
WorldState........Players.............[]*shared.Player.........!............ID.....Position......'......PlayerDisconnected........ID.................@.0..f.\.

```

Here is using json to encode message (without `omitempty` even):

```

length 101
................E...2.@.@.      ..............m.....qQ. .O...........M....K{"Sent":"2017-07-01T18:29:20.207040422-07:00","Request":null,"Update":null}
 00:00:00.074456 IP (tos 0x0, ttl 64, id 13001, offset 0, flags [DF], proto UDP (17), length 52)

```

Size of gob encoded message can be reduced if shared.Message is a interface

something like: 

```go
type Message interface{
 Encode() []byte, error
 Type() MessageType
}
```

**Comment by aerth** [comment]
From https://golang.org/pkg/encoding/gob/
> The field value is sent using the standard gob encoding for its type, recursively. If a field has the zero value for its type (except for arrays; see above), it is omitted from the transmission.

I don't believe nil pointers are omitted from the gob encoding.

**Comment by aerth** [comment]
Gob can still be used if shared.Message is a interface

something like: 
```go
type Message interface{
 Encode() []byte, error
 Type() MessageType
}
```

**Comment by ilackarms** [comment]
What's the purpose of the interface type we've defined? I ask because:

1) `Encode() ([]byte, error)` is already a global function, and putting it in an interface means that *every* kind of message needs to implement this Encode method, which will lead to a lot of duplicate code.

2) `Type() MessageType` isn't being used anywhere. Our type assertion is a `switch msg.(type)`. If we switched on`msg.Type()`, it would be more useful, but we still have to do type casting anyway (although simple type casting might be cheaper than type assertion; we should research how expensive this will be).

Personally I actually really like the current implementation, because making a `Message` an interface doesn't add any actual functionality. It's really no different than an empty `inferface{}` except it gives us some type safety (at the expense of having to duplicate code).

I think we should look into replacing `gob` to do encoding, if the only problem we're solving for is message size.
